--- a/drivers/staging/mt7621-pci/pci-mt7621.c
+++ b/drivers/staging/mt7621-pci/pci-mt7621.c
@@ -396,7 +396,9 @@
 	u8 num_disabled = 0;
 	int err;
 
+	mdelay(250);
 	mt7621_pcie_reset_assert(pcie);
+	mdelay(250);
 	mt7621_pcie_reset_rc_deassert(pcie);
 
 	list_for_each_entry_safe(port, tmp, &pcie->ports, list) {
@@ -414,6 +416,7 @@
 		}
 	}
 
+	mdelay(250);
 	mt7621_pcie_reset_ep_deassert(pcie);
 
 	tmp = NULL;
